File: julia/src/benchmarks/GraphPath.jl
using ..BenchmarkFramework
using DataStructures: BinaryMinHeap

mutable struct Graph
    vertices::Int32
    jumps::Int32
    jump_len::Int32
    adj::Vector{Vector{Int32}}

    function Graph(vertices::Int32, jumps::Int32 = 3, jump_len::Int32 = 100)
        adj = [Int32[] for _ = 1:vertices]
        new(vertices, jumps, jump_len, adj)
    end
end

function add_edge(g::Graph, u::Int64, v::Int64)
    push!(g.adj[u+1], Int32(v))
    push!(g.adj[v+1], Int32(u))
end

function generate_random(g::Graph)

    for i = 1:(g.vertices-1)
        add_edge(g, i, i-1)
    end

    for v = 0:(g.vertices-1)
        num_jumps = Helper.next_int(g.jumps)
        for _ = 1:num_jumps
            offset = Helper.next_int(g.jump_len) - g.jump_len ÷ 2
            u = v + offset

            if u >= 0 && u < g.vertices && u != v
                add_edge(g, v, u)
            end
        end
    end
end

abstract type AbstractGraphPathBenchmark <: AbstractBenchmark end

mutable struct GraphPathBFS <: AbstractGraphPathBenchmark
    graph::Graph
    result::UInt32

    function GraphPathBFS()
        vertices_val = Helper.config_i64("Graph::BFS", "vertices")
        jumps_val = Helper.config_i64("Graph::BFS", "jumps")
        jump_len_val = Helper.config_i64("Graph::BFS", "jump_len")

        graph = Graph(Int32(vertices_val), Int32(jumps_val), Int32(jump_len_val))
        new(graph, UInt32(0))
    end
end

name(b::GraphPathBFS)::String = "Graph::BFS"

function prepare(b::GraphPathBFS)
    generate_random(b.graph)
end

function bfs_shortest_path(graph::Graph, start::Int64, target::Int64)::Int32
    if start == target
        return 0
    end

    visited = falses(graph.vertices)
    queue = Vector{Tuple{Int64,Int32}}()

    visited[start+1] = true
    push!(queue, (start, Int32(0)))

    idx = 1
    while idx <= length(queue)
        v, dist = queue[idx]
        idx += 1

        for neighbor in graph.adj[v+1]
            if neighbor == target
                return dist + 1
            end

            if !visited[neighbor+1]
                visited[neighbor+1] = true
                push!(queue, (Int64(neighbor), dist + 1))
            end
        end
    end

    return -1
end

function test(b::GraphPathBFS)::Int64
    return Int64(bfs_shortest_path(b.graph, Int64(0), Int64(b.graph.vertices - 1)))
end

function run(b::GraphPathBFS, iteration_id::Int64)
    total_length = test(b)
    b.result = (b.result + UInt32(total_length)) & 0xffffffff
end

function checksum(b::GraphPathBFS)::UInt32
    return b.result
end

mutable struct GraphPathDFS <: AbstractGraphPathBenchmark
    graph::Graph
    result::UInt32

    function GraphPathDFS()
        vertices_val = Helper.config_i64("Graph::DFS", "vertices")
        jumps_val = Helper.config_i64("Graph::DFS", "jumps")
        jump_len_val = Helper.config_i64("Graph::DFS", "jump_len")

        graph = Graph(Int32(vertices_val), Int32(jumps_val), Int32(jump_len_val))
        new(graph, UInt32(0))
    end
end

name(b::GraphPathDFS)::String = "Graph::DFS"

function prepare(b::GraphPathDFS)
    generate_random(b.graph)
end

function dfs_find_path(graph::Graph, start::Int64, target::Int64)::Int32
    if start == target
        return 0
    end

    visited = falses(graph.vertices)
    stack = Vector{Tuple{Int64,Int32}}()
    best_path = typemax(Int32)

    push!(stack, (start, Int32(0)))

    while !isempty(stack)
        v, dist = pop!(stack)

        if visited[v+1] || dist >= best_path
            continue
        end
        visited[v+1] = true

        for neighbor in graph.adj[v+1]
            if neighbor == target
                if dist + 1 < best_path
                    best_path = dist + 1
                end
            elseif !visited[neighbor+1]
                push!(stack, (Int64(neighbor), dist + 1))
            end
        end
    end

    return best_path == typemax(Int32) ? -1 : best_path
end

function test(b::GraphPathDFS)::Int64
    return Int64(dfs_find_path(b.graph, Int64(0), Int64(b.graph.vertices - 1)))
end

function run(b::GraphPathDFS, iteration_id::Int64)
    total_length = test(b)
    b.result = (b.result + UInt32(total_length)) & 0xffffffff
end

function checksum(b::GraphPathDFS)::UInt32
    return b.result
end

mutable struct GraphPathAStar <: AbstractGraphPathBenchmark
    graph::Graph
    result::UInt32

    function GraphPathAStar()
        vertices_val = Helper.config_i64("Graph::AStar", "vertices")
        jumps_val = Helper.config_i64("Graph::AStar", "jumps")
        jump_len_val = Helper.config_i64("Graph::AStar", "jump_len")

        graph = Graph(Int32(vertices_val), Int32(jumps_val), Int32(jump_len_val))
        new(graph, UInt32(0))
    end
end

name(b::GraphPathAStar)::String = "Graph::AStar"

function prepare(b::GraphPathAStar)
    generate_random(b.graph)
end

function heuristic(v::Int64, target::Int64)::Int32
    return Int32(target - v)
end

function a_star_shortest_path(graph::Graph, start::Int64, target::Int64)::Int32
    if start == target
        return 0
    end

    INF = typemax(Int32)
    g_score = fill(INF, graph.vertices)
    f_score = fill(INF, graph.vertices)
    closed = falses(graph.vertices)

    g_score[start+1] = 0
    f_score[start+1] = heuristic(start, target)

    open_set = BinaryMinHeap{Tuple{Int32,Int64}}()
    in_open_set = falses(graph.vertices)

    push!(open_set, (f_score[start+1], start))
    in_open_set[start+1] = true

    while !isempty(open_set)
        f, current = pop!(open_set)
        in_open_set[current+1] = false

        if current == target
            return g_score[current+1]
        end

        closed[current+1] = true

        for neighbor in graph.adj[current+1]
            if closed[neighbor+1]
                continue
            end

            tentative_g = g_score[current+1] + 1

            if tentative_g < g_score[neighbor+1]
                g_score[neighbor+1] = tentative_g
                new_f = tentative_g + heuristic(Int64(neighbor), target)
                f_score[neighbor+1] = new_f

                if !in_open_set[neighbor+1]
                    push!(open_set, (new_f, Int64(neighbor)))
                    in_open_set[neighbor+1] = true
                end
            end
        end
    end

    return -1
end

function test(b::GraphPathAStar)::Int64
    return Int64(a_star_shortest_path(b.graph, Int64(0), Int64(b.graph.vertices - 1)))
end

function run(b::GraphPathAStar, iteration_id::Int64)
    total_length = test(b)
    b.result = (b.result + UInt32(total_length)) & 0xffffffff
end

function checksum(b::GraphPathAStar)::UInt32
    return b.result
end

File: nim/src/benchmarks/graphpath.nim
import std/[random, deques]
import ../benchmark
import ../helper

type
  Graph* = ref object
    vertices*: int
    jumps*: int
    jumpLen*: int
    adj*: seq[seq[int]]

  GraphPathBenchmark* = ref object of Benchmark
    graph*: Graph
    resultVal*: uint32

  GraphPathBFS* = ref object of GraphPathBenchmark
  GraphPathDFS* = ref object of GraphPathBenchmark
  GraphPathAStar* = ref object of GraphPathBenchmark

proc newGraph(vertices, jumps, jumpLen: int): Graph =
  result = Graph(
    vertices: vertices,
    jumps: jumps,
    jumpLen: jumpLen,
    adj: newSeq[seq[int]](vertices)
  )

proc addEdge(g: Graph, u, v: int) =
  g.adj[u].add(v)
  g.adj[v].add(u)

proc generateRandom(g: Graph) =

  for i in 1..<g.vertices:
    addEdge(g, i, i-1)

  for v in 0..<g.vertices:
    let numJumps = nextInt(int32(g.jumps))
    for _ in 0..<numJumps:
      let offset = nextInt(int32(g.jumpLen)) - int32(g.jumpLen div 2)
      let u = v + offset
      if u >= 0 and u < g.vertices and u != v:
        addEdge(g, v, u)

method prepare(self: GraphPathBenchmark) =
  let vertices = int(self.config_val("vertices"))
  let jumps = int(self.config_val("jumps"))
  let jumpLen = int(self.config_val("jump_len"))

  self.graph = newGraph(vertices, jumps, jumpLen)
  reset()
  self.graph.generateRandom()
  self.resultVal = 0

method test*(self: GraphPathBenchmark): int64 {.base.} =
  raise newException(ValueError, "Not implemented")

method run(self: GraphPathBenchmark, iteration_id: int) =
  let total = self.test()
  self.resultVal = self.resultVal + uint32(total)

method checksum(self: GraphPathBenchmark): uint32 =
  self.resultVal

proc newGraphPathBFS(): Benchmark =
  GraphPathBFS()

method name(self: GraphPathBFS): string = "GraphPathBFS"

proc bfsShortestPath(g: Graph, start, target: int): int =
  if start == target:
    return 0

  var visited = newSeq[bool](g.vertices)
  var queue = initDeque[(int, int)]()

  visited[start] = true
  queue.addLast((start, 0))

  while queue.len > 0:
    let (v, dist) = queue.popFirst()

    for neighbor in g.adj[v]:
      if neighbor == target:
        return dist + 1

      if not visited[neighbor]:
        visited[neighbor] = true
        queue.addLast((neighbor, dist + 1))

  -1

method test(self: GraphPathBFS): int64 =
  int64(bfsShortestPath(self.graph, 0, self.graph.vertices - 1))

registerBenchmark("GraphPathBFS", newGraphPathBFS)

proc newGraphPathDFS(): Benchmark =
  GraphPathDFS()

method name(self: GraphPathDFS): string = "GraphPathDFS"

proc dfsFindPath(g: Graph, start, target: int): int =
  if start == target:
    return 0

  var visited = newSeq[bool](g.vertices)
  var stack: seq[(int, int)] = @[(start, 0)]
  var bestPath = high(int)

  while stack.len > 0:
    let (v, dist) = stack.pop()

    if visited[v] or dist >= bestPath:
      continue

    visited[v] = true

    for neighbor in g.adj[v]:
      if neighbor == target:
        if dist + 1 < bestPath:
          bestPath = dist + 1
      elif not visited[neighbor]:
        stack.add((neighbor, dist + 1))

  if bestPath == high(int): -1 else: bestPath

method test(self: GraphPathDFS): int64 =
  int64(dfsFindPath(self.graph, 0, self.graph.vertices - 1))

registerBenchmark("GraphPathDFS", newGraphPathDFS)

type
  PriorityQueueItem = object
    vertex: int
    priority: int

  PriorityQueue = object
    vertices: seq[int]
    priorities: seq[int]
    size: int

proc newPriorityQueue(initialCapacity: int = 16): PriorityQueue =
  result.vertices = newSeq[int](initialCapacity)
  result.priorities = newSeq[int](initialCapacity)
  result.size = 0

proc push(pq: var PriorityQueue, vertex: int, priority: int) =
  if pq.size >= pq.vertices.len:
    let newCapacity = pq.vertices.len * 2
    pq.vertices.setLen(newCapacity)
    pq.priorities.setLen(newCapacity)

  var i = pq.size
  pq.size += 1
  pq.vertices[i] = vertex
  pq.priorities[i] = priority

  while i > 0:
    let parent = (i - 1) div 2
    if pq.priorities[parent] <= pq.priorities[i]:
      break
    swap(pq.vertices[i], pq.vertices[parent])
    swap(pq.priorities[i], pq.priorities[parent])
    i = parent

proc pop(pq: var PriorityQueue): int =
  result = pq.vertices[0]
  pq.size -= 1

  if pq.size > 0:
    pq.vertices[0] = pq.vertices[pq.size]
    pq.priorities[0] = pq.priorities[pq.size]

    var i = 0
    while true:
      let left = 2 * i + 1
      let right = 2 * i + 2
      var smallest = i

      if left < pq.size and pq.priorities[left] < pq.priorities[smallest]:
        smallest = left
      if right < pq.size and pq.priorities[right] < pq.priorities[smallest]:
        smallest = right
      if smallest == i:
        break

      swap(pq.vertices[i], pq.vertices[smallest])
      swap(pq.priorities[i], pq.priorities[smallest])
      i = smallest

proc isEmpty(pq: PriorityQueue): bool = pq.size == 0

proc newGraphPathAStar(): Benchmark =
  GraphPathAStar()

method name(self: GraphPathAStar): string = "GraphPathAStar"

proc heuristic(v, target: int): int = target - v

proc aStarShortestPath(g: Graph, start, target: int): int =
  if start == target:
    return 0

  let INF = high(int)
  var gScore = newSeq[int](g.vertices)
  var fScore = newSeq[int](g.vertices)
  var closed = newSeq[bool](g.vertices)

  for i in 0..<g.vertices:
    gScore[i] = INF
    fScore[i] = INF

  gScore[start] = 0
  fScore[start] = heuristic(start, target)

  var openSet = newPriorityQueue()
  var inOpenSet = newSeq[bool](g.vertices)

  openSet.push(start, fScore[start])
  inOpenSet[start] = true

  while not openSet.isEmpty:
    let current = openSet.pop()
    inOpenSet[current] = false

    if current == target:
      return gScore[current]

    closed[current] = true

    for neighbor in g.adj[current]:
      if closed[neighbor]:
        continue

      let tentativeG = gScore[current] + 1

      if tentativeG < gScore[neighbor]:
        gScore[neighbor] = tentativeG
        fScore[neighbor] = tentativeG + heuristic(neighbor, target)

        if not inOpenSet[neighbor]:
          openSet.push(neighbor, fScore[neighbor])
          inOpenSet[neighbor] = true

  -1

method test(self: GraphPathAStar): int64 =
  int64(aStarShortestPath(self.graph, 0, self.graph.vertices - 1))

registerBenchmark("GraphPathAStar", newGraphPathAStar)
